File: 2020/day15/solution.js
const findTurnNumber = (input, nth) => {
  const numbers = input.split(",").map(Number);

  let turn = numbers.length;
  let number = numbers.pop();
  const occurrences = new Map(numbers.map((num, idx) => [num, idx + 1]));

  while (turn < nth) {
    const next =
      turn - (occurrences.has(number) ? occurrences.get(number) : turn);
    occurrences.set(number, turn);
    number = next;
    turn += 1;
  }

  return number;
};

const input = "19,0,5,1,10,13";
console.log(findTurnNumber(input, 2020));
console.log(findTurnNumber(input, 30000000));

File: 2020/day15/solution.clj
(load-file "/src/helpers.clj")

(defn find-turn-number [input nth]
  (let [numbers (map read-string (clojure.string/split input #","))]
    (loop [turn (count numbers)
           number (last numbers)
           occurrences (transient (into {} (map-indexed (fn [idx num] [num (inc idx)]) (butlast numbers))))]
      (if (< turn nth)
        (recur (inc turn) (- turn (get occurrences number turn)) (assoc! occurrences number turn))
        number))))

(def input "19,0,5,1,10,13")
(println (find-turn-number input 2020))
(println (find-turn-number input 30000000))
